core/src/core/block.rs

```diff
@@ -212,9 +218,27 @@ impl Readable for Block {
 		let (input_len, output_len, proof_len) =
 			ser_multiread!(reader, read_u64, read_u64, read_u64);
 
-		let inputs = try!((0..input_len).map(|_| Input::read(reader)).collect());
-		let outputs = try!((0..output_len).map(|_| Output::read(reader)).collect());
-		let kernels = try!((0..proof_len).map(|_| TxKernel::read(reader)).collect());
+		let inputs: Vec<_> = try!((0..input_len).map(|_| Input::read(reader)).collect());
+		let outputs: Vec<_> = try!((0..output_len).map(|_| Output::read(reader)).collect());
+		let kernels: Vec<_> = try!((0..proof_len).map(|_| TxKernel::read(reader)).collect());
+
+		//
+		// TODO - is this the right place to do this? Is this the right approach?
+		//
+		// Consensus rule that inputs, outputs and kernels are *required* to be sorted on the wire.
+		// Simply reject anything at deserialization time if not sorted as required.
+		// This is to avoid the possibility of "fingerprinting" based on any
+		// information from the order of inputs, outputs or kernels.
+		//
+		let mut sorted_inputs = inputs.clone();
+		sorted_inputs.sort_by_key(|input| input.hash());
+		if sorted_inputs != inputs { return Err(ser::Error::BadlySorted); }
+		let mut sorted_outputs = outputs.clone();
+		sorted_outputs.sort_by_key(|output| output.hash());
+		if sorted_outputs != outputs { return Err(ser::Error::BadlySorted); }
+		let mut sorted_kernels = kernels.clone();
+		sorted_kernels.sort_by_key(|kernel| kernel.hash());
+		if sorted_kernels != kernels { return Err(ser::Error::BadlySorted); }
```

Comment:
I'm worried about the expense of the vector comparison here. It likely checks every value within each struct.

Reply:
Yeah that makes sense - let me take another look and see if we can check the order of the serialized values directly.

Reply:
Instead of the `(0..len).map` it'll likely be better to have a full blown loop that compares to the previous iteration. That can be its own little utility function.